### ISSUE TITLE
Trying to replicate the results obtained with denny brtiz's code

### ISSUE THREAD

**Issue by Quenjy** [alexander-rakhlin/CNN-for-Sentence-Classification-in-Keras#29]
First of all, thanks for sharing your code, I was looking for a keras implementation of denny britz' code because while his implementation is really well explained it's very taxing when it comes to working with bigger text corpuses, since he doesn't store the indexes of the words in each sentence but their full embedding.

However, while your code works extremely well as far as memory management is concerned, I can't get to the same results as with the denny britz's implementation

Working with my own data (pretty short and preprocessed tweets), pretrained embeddings (dimension = 300) and two classes (0 and 1) I get an accuracy of 87% on the training set and 81% on the validation set in only a few epochs with denny britz' code.

Using the code here, with the exact same layers and parameters you have set up (except for filter_sizes = [2, 3, 4, 5], which was 3 and 8 in your implementation), the performance on my data is pretty bad

![image](https://user-images.githubusercontent.com/11042499/33798168-45e5de16-dd14-11e7-82b5-6acc46d8c4b5.png)

Which makes sense, since you apparently tweaked your parameters around longer text and I'm working with tweets.

So I thought I would replicate the exact same network as denny britz to see if the performance would be comparable, here's what the code looks like now

```
filter_sizes = [2, 3, 4, 5]
num_filters = 128
hidden_dims = 20
dropout_prob = 0.5
batch_size=1024 

model_input = Input(shape=(max_sentence_len,), dtype='int32')

z =Embedding(word_embeddings.shape[0],
            word_embeddings.shape[1],
            input_length=max_sentence_len,
            weights=[word_embeddings],
            trainable=False)(model_input)



# Convolutional block
conv_blocks = []
for sz in filter_sizes:
    conv = Convolution1D(filters=num_filters,
                         filter_length=sz,
                         padding="valid", #same / valid
                         activation="relu")(z)
    conv = MaxPooling1D(pool_size=2)(conv)
    conv = Flatten()(conv)
    conv_blocks.append(conv)
z = Concatenate()(conv_blocks) if len(conv_blocks) > 1 else conv_blocks[0]

#from keras.layers import Reshape
#z = Reshape([-1, num_filters*len(filter_sizes)])(z)   #doesn't work

z = Dropout(dropout_prob)(z)
z = Dense(hidden_dims, activation="relu")(z)
model_output = Dense(1, activation="sigmoid")(z)

from keras.optimizers import Adam

#base settings : lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.0
adam_optimizer = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.0)

model = Model(model_input, model_output)
model.compile(loss="binary_crossentropy", optimizer=adam_optimizer, metrics=["accuracy"])

# Train the model
hist = model.fit(X_train, y_train, batch_size=batch_size, epochs=100,
          validation_split=0.05, verbose=2)
```

In the denny britz implementation, the training and validation accuracies increase steadily, then the validation accuracy starts going down (due to overfitting) at around 81%
With the code above, the training accuracy increases steadily up to over 90% but the validation accuracy is all over the place and never goes above 80%, it almost seems random

![image](https://user-images.githubusercontent.com/11042499/33798234-31172390-dd15-11e7-8eec-17c0e528d333.png)

Am I doing something wrong with my CNN? I can't get the reshape to work, but figured the additional hidden layer kind of did the job. I've tried tweaking its size (as well as many other parameters and layers) but can't get to the same level as the denny britz implementation. Since you most likely have way more experience with Keras than me, could you confirm whether my implementation is correct?

Or maybe the problem comes from my data itself?
Just to clarify, X_train is a tensor of the shape (rows, max_sentence_len) that contains the indexes of each word + index 0 for extra padding (index 0 leads to a vector full of 0's in the embedding file), y is an array that looks like [1 0 0 0 1 1 1 1 0 0....], max_sntences_len = 32 (since it's short tweets). I think that's the way the data is supposed to look like so I doubt the problem comes from there.

**Comment by alexander-rakhlin** [comment]
Hi,
It's hard to tell having no data. I would suggest you to 1) leave Adam settings default; set smaller lr; try other optimizer;  2) make embedding layer trainable or remove it (static mode); 3) decrease model size; 4) check weight initialization; 5) larger validation set. Apart from optimizer, which could be different implementation, I see no reasons for significant difference between the frameworks

**Comment by Quenjy** [comment]
Thanks for the quick answer,

I've played around with other optimizers and learning rates for a bit, and the same kind of thing happens everytime. The first few iterations show a normal increase, then it starts jumping up and down 5% accuracy. Here's an example of it with the code above and adagrad instead of Adam.

![image](https://user-images.githubusercontent.com/11042499/33799512-9e295a26-dd2d-11e7-858a-f8dae509b473.png)

This also happens with lower learning rates, the initial steady climb is just slower and then the jumps are less extreme (+-2% instead of 5)

Here's what my data typically looks like (before padding to size 32)

![image](https://user-images.githubusercontent.com/11042499/33799525-f4bd8de4-dd2d-11e7-99e0-354f9d3f6cf7.png)

The embeddings are the pretrained fasttext embeddings provided by facebook.

I feel like the only difference between the implementation given in my first message and the denny britz one is the reshape layer, I couldn't get it to work and figured that your way of doing it (adding one hidden layer) was kind of equivalent. Do you think that could be the issue or should I just keep playing around with hyperparameters?

**Comment by alexander-rakhlin** [comment]
Reshape layer isn't trainable and should't have an effect. Try increasing validation set size or explicitly fix this set for both frameworks.

**Comment by Quenjy** [comment]
I just increased the validation set from 10K to 40K samples (from a total of 200K labelled samples) and this is the result

![image](https://user-images.githubusercontent.com/11042499/33800422-aab7266c-dd3f-11e7-938a-a0f6f294dece.png)

I will try setting the same validation set for both this implementation and the denny britz one tomorrow, and keep you updated, it's getting pretty late here.

Thanks for taking the time to help me out anyway!

**Comment by alexander-rakhlin** [comment]
Batch size can be an issue too. Try setting to 64-128

**Comment by Quenjy** [comment]
Alright, I've set the same validation set for both denny britz's CNN and the keras implementation and I can confirm that the same trend keeps happening. The TF CNN acts as expected (both validation and training accuracy grow steadily, then at some point we overfit so the validation acc goes down while the training acc keeps going up)

But with the keras implementation, the behavior noted above keeps happening, the first few iterations work as expected but then the validation accuracy goes crazy.

I've also set the batch size to 64, but the same behavior happens

![image](https://user-images.githubusercontent.com/11042499/33804964-b3d86d02-ddb0-11e7-8d6d-dc2c3b764b99.png)

The good thing is that from time to time I reach accuracies I never reach with the denny britz CNN, the bad thing is that it looks so messy I can't really use that for a paper.


I then decided to try the keras CNN on my whole dataset (2.5 million samples) and really lowered the learning rate (100 times smaller than the default RMSprop one). This is the result

![image](https://user-images.githubusercontent.com/11042499/33807672-a67e36ee-ddda-11e7-849a-182379de647b.png)

It kind of works I guess? It's very slow but seems to steadily be going up, after 100 epochs (about 5 hours) it barely made it pas 0.8 validation accuracy though.
The weird thing is I checked and the optimizer denny uses it the same one I used at first, Adam optimizer with base parameters, so it would seem like the problem comes from the CNN and not the optimizer

**Comment by alexander-rakhlin** [comment]
Well, one last idea - turn the embedding layer trainable=True

**Comment by Quenjy** [comment]
I tried it and it didn't really change much, except for the training time. I think I will just have to keep tweaking the parameters. A low learning rate avoids the erratic validation accuracy but gets stuck in a local minima.
The funny thing is how different the results can be for learning rates that are on the same order of magnitude, for example 0.00003 gets stuck at 0.8 after 50 epochs but 0.00005 reaches 0.83 after only a few epochs but keeps jumping up and down with no clear signs of improvement.

Such is life I guess

Also I forgot to mention my keras was built on top of TF, not Theanos. But I doubt that's the problem, since incompatibility issues between the two implementations usually result in a crash and I didn't run into one.
